manifests/haproxy/autoload.pp

```diff
@@ -28,51 +32,129 @@
+      # Filter undef fields from the http-edge rules
+      $http_edge_rules_all_real_without_undef = $http_edge_rules_all.map |$key, $value| {
+        $real_rules = $value.filter |$rules| { $rules =~ NotUndef }
+        [$key, $real_rules]
+      }
+      $http_edge_rules_all_real = Hash($http_edge_rules_all_real_without_undef)
+
+      # Insert additional http-edge rules that are service specific
+      if 'canonicals' in $http_edge_services or 'status410' in $http_edge_services or or 'nofollow' in $http_edge_services {
+        if 'canonicals' in $http_edge_services {
+          $http_edge_canonicals_rules_real = flatten($http_edge_rules_all_real['canonicals'], 'set-var(txn.canonical_custom) hdr(X-Canonical-Custom) if { hdr(X-Canonical-Custom) -m found }', 'del-header X-Canonical-Custom if { hdr(X-Canonical-Custom) -m found }')
+        }
+        $http_response_rules = $ducktape::haproxy::frontends[$http_edge_frontend]['options']['http-response']
+        $http_response_new_rules = $http_edge_services.map | $service | {
+          if $service == 'canonicals' {
+            'set-header link %[var(txn.canonical_custom)] if { var(txn.canonical_custom) -m found }'
+          } elsif $service == 'status410' {
+            'set-status 410 if { var(txn.status410) -m bool }'
+          } elsif $service == 'nofollow' {
+            'set-header X-Robots-Tag "noindex, nofollow" if { var(txn.noindex) -m bool }'
           }
         }
-      })
+        $http_response_new_rules_real = $http_response_new_rules.filter |$rules| { $rules =~ NotUndef }
+        $http_response_rules_real = concat($http_response_rules, $http_response_new_rules_real)
+        $http_response_rules_hash = {$http_edge_frontend => {'options' => {'http-response' => $http_response_rules_real}}}
+      }
 
-      # Insert http-edge redirect rules at the marker position, or at the end if no marker
+      # Insert http-edge rules at the marker position, or at the end if no marker
       $http_request_rules = $ducktape::haproxy::frontends[$http_edge_frontend]['options']['http-request']
-      if grep($http_request_rules, $http_edge_redirect_marker) == [] {
-        $http_request_rules_real = concat($http_request_rules, $http_edge_redirect_rules)
+      if 'redirects' in $http_edge_services and grep($http_request_rules, $http_edge_markers['redirects']) == [] {
+        $http_request_rules_with_redirects = concat($http_request_rules, $http_edge_rules_all_real['redirects'])
+      }
+      elsif 'redirects' in $http_edge_services {
+        $http_request_rules_with_redirects = split(regsubst(join($http_request_rules, "\n"), $http_edge_markers['redirects'], join($http_edge_rules_all_real['redirects'], "\n")), "\n")
+      } else {
+        $http_request_rules_with_redirects = $http_request_rules
+      }
+
+      if 'status410' in $http_edge_services and grep($http_request_rules_with_redirects, $http_edge_markers['status410']) == [] {
```

Comment:
@jonhattan I really don't like my solution here but it is the only workaround I've found that gives the proper result considering that puppet does not allow you to reassign variables